articles/151_node_to_participant_mapping.md

```diff
@@ -0,0 +1,220 @@
+
+
+The configuration of the `Subscriber` QoS depends on how the data will be accessed later:
+- Polled using `Subscriber` read method when needed.
+- Listened and organized in a local cache.
+
+The second option allows better organization of this information (e.g.: in hash tables).
+In the first case, the QoS of the `Subscriber` should be:
+
+- Durability: Transient Local
+- History: Keep Last
+- History depth: 1
+- Reliability: Reliable
+
+In the second case, durability can be changed to `Volatile`.
+
+#### Using USER_DATA and GROUP_DATA QoSPolicy
+
+Each `Participant` could have in its user data the list of node names that owns.
+When this data is changed, each `ParticipantListener` will be notified.
+This is not a good option, as `UserData` is just a sequence of bytes.
+Organizing a complex message in it won't be easy nor performant.
+
+Similarly to `UserData`, `GroupData` is a available in `Publishers` and `Subscribers`.
+These entities only need to communicate the GUID of the `Participant` and the `Node` name from which it was created.
+This idea can be combined with a topic just publishing the list of `Node` names, without including all the other vectors in the message.
+Although, it is more difficult to communicate this information for `Services` and `Clients`, as they use behind the scenes just a `DDS Publisher` and `Subscriber`.
+
+### Other implications
+
+#### Security
+
+In `DDS`, security can be specified at a `Participant` level.
+If one `Node` is mapped to one `Participant`, individual configuration of its security key and access control policy is possible.
+From a security point of view, only being able to configure it at a `Participant` (or per process) level is enough.
+There's not much sense on having different access control policies for `Nodes` in the same process.
+As they share the same address space, other vulnerabilities are possible.
+
+##### How to create a new security key?
+
+Before, we were creating a key for each `Node`.
+The full name of the node was used for creating it.
+
+If we create one `Participant` per context, we will only need a key for each of them, and not one per `Node`.
+There are two alternatives:
+
+- Add the concept of `Context` name (or `Participant` name).
+  In this way, the key of each participant could be specified independently.
+- Use one key per process.
+  All the `Participants` within one process will use the same key.
+
+##### How to specify access policies?
+
+Access control policies could still be specified per `Node` basis.
+When a `Participant` is created, it should look at the access control policies of each of its `Nodes` and compose them in a single configuration.
```

Comment:
Is it possible to add more access control policies after creating the `Participant` (e.g.: When later creating a `Node`)?
@mikaelarguedas @ruffsl Can you review the `Security` subsection of this article?
I don't know much about `sros2` and `DDS Security`.

Reply:
> Is it possible to add more access control policies after creating the Participant [...]?

I actually don't think you can do this. @mikaelarguedas can you confirm?

Reply:
> This makes me nervous. I have a few thoughts:
> 
> * It greatly diminishes the usefulness of policies, and makes the attack surface very large. For example, any node launched in the same context as your motor controller can directly control the motors, even if it had a policy that didn't include such access.

IMO, if you don't trust in a `Node` you have to run it in another process (running something you don't trust in the same address space is insecure).
In that case, you can configure its access control policies separately.

> * Making the policy of the participant dynamic like this will make security more difficult for end users than it already is, and we're trying to _gain_ adoption!

Maybe, a better idea is to specify a policy file when running the process.
Something like `path/to/executable --ros-args __security_dir=path/to/security/dir`.
We wouldn't have `Node` level access control policies, just per process level.
In the case someone create two `contexts` in the same process, I would use the same access control policies (I don't have an use case for two contexts, so I don't have a better idea).

> * Consider nodes `A` and `B`. Node `A` has a policy that allows it to publish to topic `/a`. Node `B` has a policy that explicitly disallows access to topic `/a`. What should the combined policy contain? (I'd say refuse to combine them).

I agree.

> What is the difference between a policy that explicitly disallows access to topic `/a`, and a policy that has a default DENY and only access to topic `/b`?

If we specify access control policies per Node and combine them, I would accept access to `/a` if one file accepts it and the other defaults to deny.
But I don't like much the idea. I think specifying a file per process will be clearer to the user.

Reply:
Haven't read the spec for a while but my feeling is that permissions are passed at participant creation and not meant to be extended afterwards.
There is a notion of AccessControlListener allowing to listen to revokation of permissions during runtime, but I'm not sure this can / should be abused to extend permissions at runtime.

I don't fully agree with the idea that because there are other attack vectors in the case of running in the same process, per node access control becomes useless. Even if it's just for hygiene of development, making sure that all nodes have the exact permissions they need and no more is important. This avoids surprises down the road.

The notion of "trusting a node'  is unclear, being able to enforce a single node's interface and behavior is still very valuable. Maybe the way to cope with this is by documentation, e.g. discouraging users from using nodes in the same process until final deployment phase.

My take on merging policies:
- Explicit policies should take precedence over generic ones
- Conflicting explicit policies should result in an error
As the expected way to define policies in ROS 2s is currently a whitelist approach, conflicting policies should not happen, except if some nodes use a blacklist approach.

Reply:
> my feeling is that permissions are passed at participant creation and not meant to be extended afterwards.

Any idea how are handled things like e.g. plugins ? Assuming it requires some access that is not covered by those of the node.

Reply:
> Haven't read the spec for a while but my feeling is that permissions are passed at participant creation and not meant to be extended afterwards.
> There is a notion of AccessControlListener allowing to listen to revokation of permissions during runtime, but I'm not sure this can / should be abused to extend permissions at runtime.

Considering this and your comment about how policies should be merged, I prefer having a way of passing the access control policies to each participant (or to each executable), than merging rules specified for each `Node`.
The second option can confuse the user, and make them believe that we have access control for each node.

> I don't fully agree with the idea that because there are other attack vectors in the case of running in the same process, per node access control becomes useless. Even if it's just for hygiene of development, making sure that all nodes have the exact permissions they need and no more is important. This avoids surprises down the road.

I agree that having access control policies per Node is nice, but I think that without them we can still have secure ROS 2.
IMO, the only option for having per Node access control policies, without having a one-to-one mapping between nodes and participants, is to build an access control layer on top of DDS security, just for nodes.
I think we can do that, but I wouldn't do that in a first iteration, the amount of work for just switching from a one-to-one mapping to only one participant per context is already big.

> Any idea how are handled things like e.g. plugins ? Assuming it requires some access that is not covered by those of the node.

I think there are two options about this:
- Plugins that may be loaded are already known, so the executable is run passing sufficiently permissive access control policies.
- We add a way of dynamically loading the plugin using another participant (if we decide to have one participant per context, in another context).

Reply:
> * Plugins that may be loaded are already known, so the executable is run passing sufficiently permissive access control policies.

Well that is not always true. Anyone may develop a new plugin that the original author of the package has no idea about... Moreover appending possible plugin's access rules would be against the principle of least privilege encouraged in sros2.

Reply:
> Well that is not always true. Anyone may develop a new plugin that the original author of the package has no idea about... Moreover appending possible plugin's access rules would be against the principle of least privilege encouraged in sros2.

I wanted to say: `If plugins ...`, I've edited the original message.
I understand that that isn't always true. For the exceptions, the only solution I have in mind is the second item above (having a way of dynamically creating a new participant with different access control policies).

----

Also, I still think that though the original author doesn't know about the possible plugins that can be added, but who's running the executable may know what plugins will be used (with the exception of cases when you want to dynamically add a plugin that wasn't originally planed).
The access control policies should be written by who's running the executable, and not by the original author, if the approach we take is having access control policies per process (or per participant), and we stop supporting rules for each node.

Reply:
Sorry I was one step ahead in my previous comment, considering how to 'distribute' access policies alongside packages. That's one of the current topic in the security WG.